### PR TITLE
Support `Date32` arguments for `generate_series`

### LINKED ISSUE #9353: Support select range from datetime
### Is your feature request related to a problem or challenge?

image
duckdb did it
image

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Lordworms**
## Which issue does this PR close?

Closes #9353

### CODE REVIEW

File: datafusion/functions-array/Cargo.toml [hunk 1]
```diff
@@ -38,7 +38,8 @@ path = "src/lib.rs"
 
 [dependencies]
 arrow = { workspace = true }
-datafusion-common = { workspace = true, default-features = true }
+arrow-array = { workspace = true }
```
Comment (alamb) [inline datafusion/functions-array/Cargo.toml:41]: I wonder why we need to add arrow-array explicilty? Isn't everything we need already exported via `arrow`?

File: datafusion/sqllogictest/test_files/array.slt [hunk 2]
```diff
@@ -5549,26 +5549,32 @@ from arrays_range;
...
 select range(5),
        range(2, 5),
        range(2, 10, 3),
        range(1, 5, -1),
        range(1, -5, 1),
-       range(1, -5, -1)
+       range(1, -5, -1),
+       range(DATE '1992-09-01', DATE '1993-03-01', INTERVAL '1' MONTH),
+       range(DATE '1993-02-01', DATE '1993-01-01', INTERVAL '-1' DAY),
+       range(DATE '1989-04-01', DATE '1993-03-01', INTERVAL '1' YEAR)
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/array.slt:5561]: Can you please also add tests for null handling?

For example

```sql
       range(DATE '1992-09-01', DATE '1993-03-01', NULL,
       range(DATE '1992-09-01', NULL , INTERVAL '1' MONTH),
       range(NULL, DATE '1993-03-01', INTERVAL '1' MONTH),
```

Also, can you add negative tests (like when start is > stop by the the interval is positive (I think the code will error in this case, but it would be good to verify this

```sql
       range(DATE '1993-03-01', DATE '1989-04-01', INTERVAL '1' YEAR)
       range(, DATE '1989-04-01', DATE '1993-03-01', INTERVAL '-1' YEAR)
```

**Review by alamb — COMMENTED** [review]
Thank you @Lordworms  -- this looks great

I also tried it locally and it worked great (I found using unnest was easier to see the results)

```sql
❯ select unnest(range(DATE '1992-09-01', DATE '1993-03-01', INTERVAL '1' MONTH));
+------------------------------------------------------------------------------------------------------------+
| unnest(range(Utf8("1992-09-01"),Utf8("1993-03-01"),IntervalMonthDayNano("79228162514264337593543950336"))) |
+------------------------------------------------------------------------------------------------------------+
| 1992-09-01                                                                                                 |
| 1992-10-01                                                                                                 |
| 1992-11-01                                                                                                 |
| 1992-12-01                                                                                                 |
| 1993-01-01                                                                                                 |
| 1993-02-01                                                                                                 |
+------------------------------------------------------------------------------------------------------------+
6 rows in set. Query took 0.007 seconds.

❯ select unnest(range(DATE '1992-09-01', DATE '1993-03-01', INTERVAL '1' DAY));
+---------------------------------------------------------------------------------------------------+
| unnest(range(Utf8("1992-09-01"),Utf8("1993-03-01"),IntervalMonthDayNano("18446744073709551616"))) |
+---------------------------------------------------------------------------------------------------+
| 1992-09-01                                                                                        |
| 1992-09-02                                                                                        |
| 1992-09-03                                                                                        |
| 1992-09-04                                                                                        |
...
```
I think it just needs a few more tests and this PR will be ready to go

File: datafusion/sqllogictest/test_files/array.slt (hunk 2 above)
Comment (Lordworms) [reply datafusion/sqllogictest/test_files/array.slt:5561]: OK, I will do it now

File: datafusion/functions-array/Cargo.toml (hunk 1 above)
Comment (Lordworms) [reply datafusion/functions-array/Cargo.toml:41]: I will remove it. I was trying to use some methods in that crate so imported that, I will remove it

**Comment by alamb** [comment]
Bonus points for updating the docs to mention the new support: https://arrow.apache.org/datafusion/user-guide/sql/scalar_functions.html#range

(we can do this as a follow on PR as well)

**Comment by Lordworms** [comment]
> Bonus points for updating the docs to mention the new support for date: https://arrow.apache.org/datafusion/user-guide/sql/scalar_functions.html#range
> 
> (we can do this as a follow on PR as well)

Yes, I will do it in next PR, along with finishing port of regx function today

**Review by alamb — APPROVED** [review]
Thank you @Lordworms

**Comment by alamb** [comment]
Screenshot 2024-03-03 at 3 38 15 AM

🤔  looks like there are some conflicts to resolve

**Comment by Lordworms** [comment]
> Screenshot 2024-03-03 at 3 38 15 AM
> 🤔 looks like there are some conflicts to resolve

got it

**Review by alamb — APPROVED** [review]
Thanks again @Lordworms
